File: LeetCode/Explore/BinTree/PreOrderTraversal.cpp
#include "../main.h"

class Solution {
  public:
    vector<int> preorderTraversal(TreeNode *root) {
        vector<int> res;
        pre(root, res);
        return res;
    }

  private:
    void pre(TreeNode *root, vector<int> &res) {
        if (root == nullptr)
            return;
        res.push_back(root->val);
        if (root->left)
            pre(root->left, res);
        if (root->right)
            pre(root->right, res);
    }
};

File: LeetCode/Explore/binary-tree/PreOrderTraversal.go
package bintree

func preorderTraversal(root *TreeNode) []int {
	var res = make([]int, 0)
	preorder(root, &res)
	return res
}

func preorder(root *TreeNode, res *[]int) {
	if root == nil {
		return
	}
	*res = append(*res, root.Val)
	preorder(root.Left, res)
	preorder(root.Right, res)
}
